File: app/src/main/java/com/rokudo/xpense/data/repositories/TransactionRepo.java
package com.rokudo.xpense.data.repositories;

import android.util.Log;

import androidx.annotation.Nullable;
import androidx.lifecycle.MutableLiveData;

import com.google.android.gms.common.util.CollectionUtils;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.EventListener;
import com.google.firebase.firestore.FirebaseFirestoreException;
import com.google.firebase.firestore.ListenerRegistration;
import com.google.firebase.firestore.Query;
import com.google.firebase.firestore.QuerySnapshot;
import com.rokudo.xpense.models.Transaction;
import com.rokudo.xpense.utils.DatabaseUtils;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.Collection;
import java.util.Date;
import java.util.List;

public class TransactionRepo {
    private static final String TAG = "TransactionRepo";

    public static TransactionRepo instance;

    private ListenerRegistration transactionListener;
    private ListenerRegistration latestTransListener;

    private final MutableLiveData<List<Transaction>> allTransactionList;
    private final List<Transaction> storedTransactionList;
    private final MutableLiveData<Transaction> latestTransaction;
    private final MutableLiveData<String> addTransactionStatus;
    private final MutableLiveData<String> updateTransactionStatus;

    private String storedWalletId;

    public static TransactionRepo getInstance() {
        if (instance == null) {
            instance = new TransactionRepo();
        }
        return instance;
    }

    public TransactionRepo() {
        this.allTransactionList = new MutableLiveData<>();
        this.storedTransactionList = new ArrayList<>();
        this.addTransactionStatus = new MutableLiveData<>();
        this.latestTransaction = new MutableLiveData<>();
        this.updateTransactionStatus = new MutableLiveData<>();
    }

    public void removeAllTransactionsData() {
        if (transactionListener != null) {
            transactionListener.remove();
        }
        if (latestTransListener != null) {
            latestTransListener.remove();
        }
        allTransactionList.setValue(null);
        storedTransactionList.clear();
        addTransactionStatus.setValue(null);
        latestTransaction.setValue(null);
        updateTransactionStatus.setValue(null);
    }

    public MutableLiveData<List<Transaction>> loadTransactions() {
        return allTransactionList;
    }

    public MutableLiveData<List<Transaction>> loadTransactions(String walletId, Date startDate) {
        Calendar calendar = Calendar.getInstance();
        calendar.set(calendar.get(Calendar.YEAR),
                calendar.get(Calendar.MONTH),
                calendar.getActualMaximum(Calendar.DAY_OF_MONTH),
                23, 59);
        Date end = calendar.getTime();

        if (transactionListener != null) {
            transactionListener.remove();
        }
        transactionListener = DatabaseUtils.getTransactionsRef(walletId)
                .whereGreaterThan("date", startDate)
                .whereLessThan("date", end)
                .orderBy("date", Query.Direction.DESCENDING)
                .addSnapshotListener((value, error) -> {
                    if (error != null || value == null || value.isEmpty()) {
                        Log.e(TAG, "loadTransactions: null or error: ", error);
                        allTransactionList.setValue(new ArrayList<>());
                        storedTransactionList.clear();
                        latestTransaction.setValue(new Transaction());
                    } else {
                        List<Transaction> transactionList = new ArrayList<>();
                        for (DocumentSnapshot documentSnapshot : value) {
                            Transaction transaction = documentSnapshot.toObject(Transaction.class);
                            if (transaction != null) {
                                transaction.setId(documentSnapshot.getId());
                                transactionList.add(transaction);
                            }
                        }
                        latestTransaction.setValue(transactionList.size() == 0 ? new Transaction() : transactionList.get(0));
                        allTransactionList.setValue(transactionList);
                        storedTransactionList.clear();
                        storedTransactionList.addAll(transactionList);
                    }
                });

        return allTransactionList;
    }

    public MutableLiveData<String> addTransaction(Transaction transaction) {

        DatabaseUtils.getTransactionsRef(transaction.getWalletId()).document(transaction.getId())
                .set(transaction)
                .addOnSuccessListener(result -> addTransactionStatus.setValue("Success"));

        return addTransactionStatus;
    }

    public MutableLiveData<Transaction> loadLatestTransaction(String walletId) {
        if (latestTransListener == null || storedWalletId == null || !storedWalletId.equals(walletId)) {
            if (latestTransListener != null) {
                latestTransListener.remove();
            }

            storedWalletId = walletId;

            latestTransListener = DatabaseUtils.getTransactionsRef(walletId)
                    .orderBy("date", Query.Direction.DESCENDING)
                    .limit(1)
                    .addSnapshotListener((value, error) -> {
                        if (value == null || value.isEmpty()) {
                            return;
                        }
                        Transaction transaction = value.getDocuments().get(0).toObject(Transaction.class);
                        if (transaction == null) {
                            return;
                        }
                        latestTransaction.setValue(transaction);
                    });
        }
        return latestTransaction;
    }

    public MutableLiveData<List<Transaction>> loadTransactionsDateInterval(String walletId, Date start, Date end) {
        MutableLiveData<List<Transaction>> data = new MutableLiveData<>();
        DatabaseUtils.getTransactionsRef(walletId)
                .whereGreaterThan("date", start)
                .whereLessThan("date", end)
                .orderBy("date", Query.Direction.DESCENDING)
                .get().addOnSuccessListener(value -> {
                    if (value == null || value.isEmpty()) {
                        data.setValue(new ArrayList<>());
                    } else {
                        List<Transaction> transactionList = new ArrayList<>();
                        for (DocumentSnapshot documentSnapshot : value) {
                            Transaction transaction = documentSnapshot.toObject(Transaction.class);
                            if (transaction != null) {
                                transaction.setId(documentSnapshot.getId());
                                transactionList.add(transaction);
                            }
                        }
                        data.setValue(transactionList);
                    }
                });

        return data;
    }

    public List<Transaction> getStoredTransactionList() {
        return storedTransactionList;
    }

    public MutableLiveData<String> updateTransaction(Transaction transaction) {
        DatabaseUtils.getTransactionsRef(transaction.getWalletId()).document(transaction.getId())
                .set(transaction)
                .addOnSuccessListener(result -> updateTransactionStatus.setValue("Success"));

        return updateTransactionStatus;
    }

    public MutableLiveData<Boolean> deleteTransaction(String transactionId, String walletId) {
        MutableLiveData<Boolean> result = new MutableLiveData<>();

        DatabaseUtils.getTransactionsRef(walletId)
                .document(transactionId)
                .delete()
                .addOnCompleteListener(task -> result.setValue(task.isSuccessful()));

        return result;
    }
}

File: app/src/main/java/com/rokudo/xpense/data/repositories/TransactionRepo.kt
package com.rokudo.xpense.data.repositories

import android.util.Log
import androidx.lifecycle.MutableLiveData
import com.rokudo.xpense.data.api.ApiServiceProvider
import com.rokudo.xpense.models.Transaction
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.util.*

class TransactionRepo {
    companion object {
        private const val TAG = "TransactionRepo"
        val instance by lazy { TransactionRepo() }
    }

    private val allTransactionList: MutableLiveData<List<Transaction>?> = MutableLiveData<List<Transaction>?>()
    private val storedTransactionList = mutableListOf<Transaction>()
    private val latestTransaction: MutableLiveData<Transaction?> = MutableLiveData<Transaction?>()
    private val addTransactionStatus: MutableLiveData<String?> = MutableLiveData<String?>()
    private val updateTransactionStatus: MutableLiveData<String?> = MutableLiveData<String?>()
    private var storedWalletId: String? = null

    fun removeAllTransactionsData() {
        allTransactionList.value = null
        storedTransactionList.clear()
        addTransactionStatus.value = null
        latestTransaction.value = null
        updateTransactionStatus.value = null
    }

    fun loadTransactions(): MutableLiveData<List<Transaction>?> = allTransactionList

    fun loadTransactions(walletId: String, startDate: Date): MutableLiveData<List<Transaction>?> {
        val calendar = Calendar.getInstance().apply {
            set(get(Calendar.YEAR), get(Calendar.MONTH), getActualMaximum(Calendar.DAY_OF_MONTH), 23, 59)
        }
        val end = calendar.time

        CoroutineScope(Dispatchers.IO).launch {
            try {
                val response = ApiServiceProvider.transactionService.getWalletTransactions(walletId)
                if (response.isSuccessful) {
                    val transactions = response.body() ?: emptyList()
                    // Filter by date range client-side
                    val filtered = transactions.filter { t ->
                        val date = t.date
                        date != null && date.after(startDate) && date.before(end)
                    }.sortedByDescending { it.date }

                    latestTransaction.postValue(if (filtered.isEmpty()) Transaction() else filtered[0])
                    allTransactionList.postValue(filtered)
                    storedTransactionList.clear()
                    storedTransactionList.addAll(filtered)
                } else {
                    Log.e(TAG, "loadTransactions: ${response.errorBody()?.string()}")
                    allTransactionList.postValue(emptyList())
                    storedTransactionList.clear()
                    latestTransaction.postValue(Transaction())
                }
            } catch (e: Exception) {
                Log.e(TAG, "loadTransactions: ${e.message}", e)
                allTransactionList.postValue(emptyList())
                storedTransactionList.clear()
                latestTransaction.postValue(Transaction())
            }
        }
        return allTransactionList
    }

    fun addTransaction(transaction: Transaction): MutableLiveData<String?> {
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val response = ApiServiceProvider.transactionService.createTransaction(
                    transaction.walletId ?: "",
                    transaction
                )
                if (response.isSuccessful) {
                    addTransactionStatus.postValue("Success")
                } else {
                    val error = response.errorBody()?.string() ?: "Creation failed"
                    Log.e(TAG, "addTransaction: $error")
                    addTransactionStatus.postValue(error)
                }
            } catch (e: Exception) {
                Log.e(TAG, "addTransaction: ${e.message}", e)
                addTransactionStatus.postValue(e.message)
            }
        }
        return addTransactionStatus
    }

    fun loadLatestTransaction(walletId: String): MutableLiveData<Transaction?> {
        if (storedWalletId == null || storedWalletId != walletId) {
            storedWalletId = walletId
            fetchLatestTransaction(walletId)
        }
        return latestTransaction
    }

    /**
     * Force refresh all transaction data for the current wallet.
     * Call this after adding/updating/deleting a transaction.
     */
    fun refreshData(walletId: String) {
        storedWalletId = walletId
        fetchLatestTransaction(walletId)
    }

    private fun fetchLatestTransaction(walletId: String) {
        CoroutineScope(Dispatchers.IO).launch {
            try {
                // Try the dedicated /latest endpoint first
                val response = ApiServiceProvider.statisticsService.getLatestTransaction(walletId)
                if (response.isSuccessful) {
                    latestTransaction.postValue(response.body())
                } else {
                    // Fallback: fetch all and pick the latest
                    val fallback = ApiServiceProvider.transactionService.getWalletTransactions(walletId)
                    if (fallback.isSuccessful) {
                        val transactions = fallback.body() ?: emptyList()
                        val latest = transactions.maxByOrNull { it.date ?: Date(0) }
                        latestTransaction.postValue(latest)
                    }
                }
            } catch (e: Exception) {
                Log.e(TAG, "fetchLatestTransaction: ${e.message}", e)
            }
        }
    }

    fun loadTransactionsDateInterval(walletId: String, start: Date, end: Date): MutableLiveData<List<Transaction>> {
        val data = MutableLiveData<List<Transaction>>()
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val response = ApiServiceProvider.transactionService.getWalletTransactions(walletId)
                if (response.isSuccessful) {
                    val transactions = response.body() ?: emptyList()
                    val filtered = transactions.filter { t ->
                        val date = t.date
                        date != null && date.after(start) && date.before(end)
                    }.sortedByDescending { it.date }
                    data.postValue(filtered)
                } else {
                    data.postValue(emptyList())
                }
            } catch (e: Exception) {
                Log.e(TAG, "loadTransactionsDateInterval: ${e.message}", e)
                data.postValue(emptyList())
            }
        }
        return data
    }

    fun getStoredTransactionList(): List<Transaction> = storedTransactionList

    /**
     * Load the N most recent transactions using the dedicated /recent endpoint.
     */
    fun loadRecentTransactions(walletId: String, limit: Int = 5): MutableLiveData<List<Transaction>> {
        val data = MutableLiveData<List<Transaction>>()
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val response = ApiServiceProvider.statisticsService.getRecentTransactions(walletId, limit)
                if (response.isSuccessful) {
                    data.postValue(response.body() ?: emptyList())
                } else {
                    // Fallback: fetch all and take the most recent
                    val fallback = ApiServiceProvider.transactionService.getWalletTransactions(walletId)
                    if (fallback.isSuccessful) {
                        val recent = (fallback.body() ?: emptyList())
                            .sortedByDescending { it.date }
                            .take(limit)
                        data.postValue(recent)
                    } else {
                        data.postValue(emptyList())
                    }
                }
            } catch (e: Exception) {
                Log.e(TAG, "loadRecentTransactions: ${e.message}", e)
                data.postValue(emptyList())
            }
        }
        return data
    }

    fun updateTransaction(transaction: Transaction): MutableLiveData<String?> {
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val response = ApiServiceProvider.transactionService.updateTransaction(
                    transaction.walletId ?: "",
                    transaction.id ?: "",
                    transaction
                )
                if (response.isSuccessful) {
                    updateTransactionStatus.postValue("Success")
                } else {
                    val error = response.errorBody()?.string() ?: "Update failed"
                    Log.e(TAG, "updateTransaction: $error")
                    updateTransactionStatus.postValue(error)
                }
            } catch (e: Exception) {
                Log.e(TAG, "updateTransaction: ${e.message}", e)
                updateTransactionStatus.postValue(e.message)
            }
        }
        return updateTransactionStatus
    }

    fun deleteTransaction(transactionId: String, walletId: String): MutableLiveData<Boolean> {
        val result = MutableLiveData<Boolean>()
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val response = ApiServiceProvider.transactionService.deleteTransaction(walletId, transactionId)
                result.postValue(response.isSuccessful)
            } catch (e: Exception) {
                Log.e(TAG, "deleteTransaction: ${e.message}", e)
                result.postValue(false)
            }
        }
        return result
    }
}
